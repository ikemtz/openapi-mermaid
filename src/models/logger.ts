export interface ILogger {
  log: (data: string) => void;
  error: (data: string) => void;
  warn: (data: string) => void;
}
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-empty-function */
export class MockConsoleLogger implements ILogger {
  public log = (data: string) => {};
  public error = (data: string) => {};
  public warn = (data: string) => {};
}
