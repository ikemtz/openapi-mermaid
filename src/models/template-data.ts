export interface ITemplateData {
  entities?: IEntity[];
  paths: string[];
}
export interface IEntity {
  isEnum?: boolean;
  enumValues: (string | { key?: number; name: string })[];
  name: string;
  importTypes: IImportType[];
  valueProperties: IValueProperty[];
  referenceProperties: IReferenceProperty[];
}
export interface IImportType {
  name: string;
}
export interface IValueProperty {
  name: string;
  type?: string;
  isArray: boolean;
  hasMultipleValidators: boolean;
  required: boolean;
  maxLength?: number;
  minLength?: number;
}

export interface IReferenceProperty {
  name: string;
  referenceTypeName: string;
  isArray: boolean;
  required: boolean;
}
