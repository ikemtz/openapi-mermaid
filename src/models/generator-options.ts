/* eslint-disable @typescript-eslint/no-unused-vars */
import { resolve } from 'path';
import { ILogger } from './logger';
import { IEntity, IReferenceProperty, IValueProperty } from './template-data';

export interface IGeneratorOptions {
  logger?: ILogger;
  outputPath: string;
  outputFileName?: string;
  openApiJsonUrl?: string;
  openApiJsonFileName?: string;
  typeFilterCallBack?: (entity: IEntity, index: number, array: IEntity[]) => boolean;
  valuePropertyTypeFilterCallBack?: (valueProperty: IValueProperty, index: number, array: IValueProperty[]) => boolean;
  referencePropertyTypeFilterCallBack?: (referenceProperty: IReferenceProperty, index: number, array: IReferenceProperty[]) => boolean;
  templates?: ITemplates | null;
}
export interface ITemplates {
  htmlClassDiagram?: string;
  mdClassDiagram?: string;
}
export function defaultFilter(
  value: IEntity | IValueProperty | IReferenceProperty,
  index: number,
  array: IEntity[] | IValueProperty[] | IReferenceProperty[],
): boolean {
  return true;
}

export function setGeneratorOptionDefaults(options: IGeneratorOptions): IGeneratorOptions {
  const templateFolder = resolve(`${__dirname}/..`, 'templates');
  options.typeFilterCallBack = options.typeFilterCallBack ?? defaultFilter;
  options.valuePropertyTypeFilterCallBack = options.valuePropertyTypeFilterCallBack ?? defaultFilter;
  options.referencePropertyTypeFilterCallBack = options.referencePropertyTypeFilterCallBack ?? defaultFilter;
  options.templates = {
    ...options.templates,
    mdClassDiagram: options.templates?.mdClassDiagram ?? `${templateFolder}/class-diagram.md.hbs`,
    htmlClassDiagram: options.templates?.htmlClassDiagram ?? `${templateFolder}/class-diagram.html.hbs`,
  };
  return options;
}
