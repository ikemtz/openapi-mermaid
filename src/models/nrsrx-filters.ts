/* eslint-disable @typescript-eslint/no-unused-vars */
import { IEntity, IValueProperty } from './template-data';

export function nrsrxTypeFilterCallBack(val: IEntity, i: number, arr: IEntity[]): boolean {
  return !val.name.endsWith('ODataEnvelope');
}
export function nrsrxValuePropertyTypeFilterCallBack(val: IValueProperty, i: number, arr: IValueProperty[]): boolean {
  return !val.name.startsWith('created') && !val.name.startsWith('updated');
}
