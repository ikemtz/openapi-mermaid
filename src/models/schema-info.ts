import { ReferenceObject, SchemaObject } from 'openapi3-ts';
import { defaultFilter, IGeneratorOptions } from './generator-options';
import { IReferenceProperty, IValueProperty } from './template-data';

export class SchemaWrapperInfo {
  public propertySchemaObject: SchemaObject = {};
  public propertyReferenceObject: ReferenceObject = { $ref: '' };

  public readonly componentSchemaObject: SchemaObject;
  public readonly valueProperties: IValueProperty[];
  public referenceProperties: IReferenceProperty[];

  constructor(schemaItem: SchemaObject) {
    this.componentSchemaObject = schemaItem;
    this.valueProperties = [];
    this.referenceProperties = [];
  }

  public updateReferenceProperties(options: IGeneratorOptions): void {
    this.referenceProperties = this.referenceProperties.filter(options.referencePropertyTypeFilterCallBack || defaultFilter);
  }
}
