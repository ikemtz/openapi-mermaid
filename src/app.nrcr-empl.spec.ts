import { mkdirSync } from 'fs';
import { generateDiagrams, nrsrxTypeFilterCallBack, nrsrxValuePropertyTypeFilterCallBack } from '.';
import { ValidateFiles } from './app.spec';
import { IGeneratorOptions } from './models/generator-options';

const fileEmployeeGenerationOptionsFactory = (): IGeneratorOptions => ({
  openApiJsonFileName: '../src/open-api-spec-docs/nrcrn-empl.json',
  outputPath: './jest_output/empl/',
  typeFilterCallBack: nrsrxTypeFilterCallBack,
  valuePropertyTypeFilterCallBack: nrsrxValuePropertyTypeFilterCallBack,
});

describe('File Based - Full Integration Tests', () => {
  describe('NRSRx Employee OData Microservice', () => {
    const options = fileEmployeeGenerationOptionsFactory();
    it('should generate files', async () => {
      try {
        mkdirSync(options.outputPath, { recursive: true });
      } catch {
        // ignore
      }
      await generateDiagrams(options);
      ValidateFiles(options);
    });
  });
});
