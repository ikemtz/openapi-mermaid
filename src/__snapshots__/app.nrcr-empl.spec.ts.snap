// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Based - Full Integration Tests NRSRx Employee OData Microservice should generate files 1`] = `
Array [
  "class-diagram.md",
]
`;

exports[`File Based - Full Integration Tests NRSRx Employee OData Microservice should generate files: class-diagram.md 1`] = `
"\`\`\`mermaid
classDiagram

  Employee ..> EmployeeCertification
  Employee ..> EmployeeCompetency
  Employee ..> EmployeeHealthItem

  class EmployeeCertification{
    +uuid: id
    +uuid: employeeId
    +uuid: certificationId
    +string: certificationName
    +date: expiresOnUtc
  }
  class EmployeeCompetency{
    +uuid: id
    +uuid: employeeId
    +uuid: competencyId
    +string: competencyName
    +date: expiresOnUtc
    +bool: isEnabled
  }
  class EmployeeHealthItem{
    +uuid: id
    +uuid: employeeId
    +uuid: healthItemId
    +string: healthItemName
    +date: expiresOnUtc
    +bool: isEnabled
  }
  class Employee{
    +uuid: id
    +string: lastName
    +string: firstName
    +date: birthDate
    +string: mobilePhone
    +string: homePhone
    +string: photo
    +string: email
    +string: addressLine1
    +string: addressLine2
    +string: city
    +string: state
    +string: zip
    +bool: isEnabled
    +date: hireDate
    +date: fireDate
    +number: totalHoursOfService
    +number: certificationCount
    +number: competencyCount
    +number: healthItemCount
    +EmployeeCertification[]: employeeCertifications;
    +EmployeeCompetency[]: employeeCompetencies;
    +EmployeeHealthItem[]: employeeHealthItems;
  }
\`\`\`
This file was generated by the [openapi-mermaid](https://www.npmjs.com/package/openapi-mermaid).
"
`;
